Fix issue 13661

yeswanth-s1th opened the pull request:
closes issue [#13661](https://github.com/apache/iceberg/issues/13661)
docs: improve local editing experience

This commit introduces a new `serve-local` command to the Makefile that allows for a much faster local build of the documentation.

The `serve-local` command works by creating symbolic links to the local `docs` and `javadoc` directories instead of pulling all the versioned documentation from the remote repository. This significantly reduces the build time, making it much easier to edit and preview the documentation locally.

The following changes were made:

- A new `pull_local_docs` function was added to `site/dev/common.sh` that creates symbolic links to the local `docs` and `javadoc` directories.
- The `site/dev/setup_env.sh` script was modified to include a `--local` flag that triggers the use of the `pull_local_docs` function.
- The `site/dev/serve.sh` script was modified to accept and pass on the `--local` flag.
- A new `serve-local` command was added to the `site/Makefile` that executes `serve.sh` with the `--local` flag.

kevinjqliu commented:
thank you for working on this! could you also try to benchmark how long `make serve` takes before and after this change?
Speed is one of the primary motivation for #13661

yeswanth-s1th commented:
Hi @kevinjqliu, 
I’ve measured the cold‑start latency of `make serve` before vs. after this patch using a simple start‑to‑banner timer:

Ref | Startup Time
-- | --
main | 733 s
PR #13664 | 345 s
Δ | ≈ 53 % faster


These results demonstrate a ~53 % acceleration in our docs‑serve delivery pipeline, significantly reducing the turnaround for iterative doc previews. Tested on macOS (M1 Pro, Python 3.8.2). No functional regressions observed.

Let me know if you’d like any additional profiling data or comparisons on other platforms. 
Thanks for reviewing!

yeswanth-s1th commented:
Hi @kevinjqliu 👋

Thanks for reviewing this enhancement. Unless you have any blocking concerns, I’m ready for approval and merge. Would you mind giving this PR a 👍 and approving/merging it when you have a moment?

Appreciate your time! 🚀

yeswanth-s1th commented:
Hi @electrum / @kevinjqliu / @martint, 
Would you mind reviewing this enhancement. Unless you have any blocking concerns, I’m ready for approval and merge. Would you mind giving this PR a 👍 and approving/merging it when you have a moment?

Appreciate your time!

manuzhang commented on site/dev/common.sh:
```diff
@@ -201,7 +201,9 @@ search_exclude_versioned_docs () {
   # Modify .md files to exclude versioned documentation from search indexing
   python3 -c "import os
-for f in filter(lambda x: x.endswith('.md'), os.listdir()): lines = open(f).readlines(); open(f, 'w').writelines(lines[:2] + ['search:\n', '  exclude: true\n'] + lines[2:]);"
+for f in filter(lambda x: x.endswith('.md'), os.listdir()): lines = open(f).readlines(); open(f, 'w').writelines(lines[:2] + ['search:
```
what's changed here?

manuzhang commented on site/Makefile:
```diff
@@ -15,12 +15,17 @@
 
 .PHONY: help
 help: # Show help for each of the Makefile recipes.
```
why is this changed?

manuzhang commented on site/dev/common.sh:
```diff
@@ -174,17 +174,18 @@ update_version () {
 
-  # Update version information within the mkdocs.yml file using sed commands
-  if [ "$(uname)" == "Darwin" ]
+    # Update version information within the mkdocs.yml file using sed commands
```
Nit: extra indent

yeswanth-s1th replied:
fixed it, removed extra indent

manuzhang commented on site/dev/common.sh:
```diff
@@ -174,17 +174,18 @@ update_version () {
-  # Update version information within the mkdocs.yml file using sed commands
-  if [ "$(uname)" == "Darwin" ]
+    # Update version information within the mkdocs.yml file using sed commands
+  if [ "$(uname)" = "Darwin" ]; 
```
do we need to change this part?

yeswanth-s1th replied:
not necessary, fixed it

yeswanth-s1th commented:
> @manuzhang 
> 
> Thanks for the thorough review. All the comments raised so far were formatting/stray edits on my side (extra ;, accidental ESC chars, duplicate sed lines, etc.). I’ve cleaned those up and reverted the noise so the PR now only contains the intended local-docs changes. Sorry for the unnecessary churn earlier, and appreciate you catching them.
> 
> 
> 
> Could you please take another look and let me know if it’s good to approve?

@manuzhang Could you please take another look and let me know if it’s good to approve?

manuzhang commented on site/dev/common.sh:
```diff
@@ -177,14 +177,15 @@ update_version () {
-    /usr/bin/sed -i '' -E "s/(^site\_name:[[:space:]]+docs\/).*$/\1${ICEBERG_VERSION}/" ${ICEBERG_VERSION}/mkdocs.yml
-    /usr/bin/sed -i '' -E "s/(^[[:space:]]*-[[:space:]]+Javadoc:.*\/javadoc\/).*$/\1${ICEBERG_VERSION}/" ${ICEBERG_VERSION}/mkdocs.yml
-  elif [ "$(expr substr $(uname -s) 1 5)" == "Linux" ]
+    /usr/bin/sed -i '' -E "s/(^site_name:[[:space:]]+docs\/).*$/\1${ICEBERG_VERSION}/" "${ICEBERG_VERSION}/mkdocs.yml"
```
@yeswanth-s1th have you checked this?

yeswanth-s1th replied:
Thanks for following up. Yes, I’ve rechecked this.

On Linux, I kept the original token-level pattern ([^[:space:]]+) so it only replaces the version number after docs/ without affecting anything else on the line. On macOS and for the Javadoc lines, I left them as they are in main (.*$) so their behavior doesn’t change.

I also added quotes around $(uname -s) for safety. This keeps the behavior consistent with main while addressing the earlier concern, and keeps the change minimal.

manuzhang replied:
But I don't think the change is related to this PR. I'd prefer having the enhancement in a separate PR.

yeswanth-s1th replied:
@manuzhang 
Makes sense. I’ve reverted this block to match main so the PR stays focused. I’ll open a separate PR for the small sed/quoting cleanups.

yeswanth-s1th commented:
> > @manuzhang 
> > 
> > Thanks for the thorough review. All the comments raised so far were formatting/stray edits on my side (extra ;, accidental ESC chars, duplicate sed lines, etc.). I’ve cleaned those up and reverted the noise so the PR now only contains the intended local-docs changes. Sorry for the unnecessary churn earlier, and appreciate you catching them.
> > 
> > 
> > 
> > Could you please take another look and let me know if it’s good to approve?
> 
> @manuzhang Could you please take another look and let me know if it’s good to approve?

@manuzhang 

Just following up — Could you take another look when you get a chance? Let me know if anything else needs to be addressed. Thanks!

manuzhang commented:
@yeswanth-s1th have you checked my latest comments?
